lib/c-api/src/wasm_c_api/types/extern_.rs

```diff
@@ -75,9 +36,81 @@ impl From<&ExternType> for wasm_externkind_enum {
     }
 }
 
+#[derive(Debug, Clone)]
+pub(crate) enum WasmExternType {
+    Function(WasmFunctionType),
+    Global(WasmGlobalType),
+    Table(WasmTableType),
+    Memory(WasmMemoryType),
+}
+
+#[allow(non_camel_case_types)]
+#[derive(Debug, Clone)]
+pub struct wasm_externtype_t {
+    pub(crate) inner: WasmExternType,
+}
+
+impl wasm_externtype_t {
+    pub(crate) fn new(extern_type: ExternType) -> Self {
+        Self {
+            inner: match extern_type {
+                ExternType::Function(function_type) => {
+                    WasmExternType::Function(WasmFunctionType::new(function_type))
+                }
+                ExternType::Global(global_type) => {
+                    WasmExternType::Global(WasmGlobalType::new(global_type))
+                }
+                ExternType::Table(table_type) => {
+                    WasmExternType::Table(WasmTableType::new(table_type))
+                }
+                ExternType::Memory(memory_type) => {
+                    WasmExternType::Memory(WasmMemoryType::new(memory_type))
+                }
+            },
+        }
+    }
+}
+
+impl From<ExternType> for wasm_externtype_t {
+    fn from(extern_type: ExternType) -> Self {
+        Self::new(extern_type)
+    }
+}
+
+impl From<&ExternType> for wasm_externtype_t {
+    fn from(other: &ExternType) -> Self {
+        other.clone().into()
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasm_extern_type(r#extern: &wasm_extern_t) -> Box<wasm_externtype_t> {
+    Box::new(wasm_externtype_t::new(r#extern.inner.ty()))
+}
+
 #[no_mangle]
-pub unsafe extern "C" fn wasm_externtype_kind(et: &wasm_externtype_t) -> wasm_externkind_t {
-    wasm_externkind_enum::from(&et.inner) as wasm_externkind_t
+pub unsafe extern "C" fn wasm_extern_kind(r#extern: &wasm_extern_t) -> wasm_externkind_t {
+    wasm_externkind_enum::from(r#extern.inner.ty()) as wasm_externkind_t
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasm_externtype_delete(_et: Option<Box<wasm_externtype_t>>) {}
```

Comment:
Optional: FWIW I'd just call it `_` in this case.

Reply:
I like being super explicit. In this case, I would rename it to `_extern_type` actually :-D.